### PR TITLE
Add MSSQLRawQuery block

### PR DESCRIPTION
**Pull request by deliciousmonster**

### CODE REVIEW

**Review by mattdodge — CHANGES_REQUESTED** [review]
I'm ok with having a general query block, but we should not require the user to enter SQL every time they want to insert/query/etc. So we should not remove the old blocks, this would be just a new block they could use if needed.

**Comment by deliciousmonster** [comment]
> I'm ok with having a general query block, but we should not require the user to enter SQL every time they want to insert/query/etc. So we should not remove the old blocks, this would be just a new block they could use if needed.

This is complete. I also cleaned up the spec and readme

File: mssql_base.py [hunk 1]
```diff
@@ -1,91 +1,85 @@
...
-@not_discoverable
-class MSSQLBase(Block):
-
-    version = VersionProperty('0.1.0')
-    server = StringProperty(title='Server', order=1)
-    port = IntProperty(title='Port', default=1433, order=2)
-    database = StringProperty(title='Database', order=3)
-    credentials = ObjectProperty(Credentials, title='Connection Credentials',
-                                 order=4)
-    table = StringProperty(title='Table', default='{{ $table }}', order=5)
```
Comment (32bitkid) [inline mssql_base.py:23]: rather than remove the `table` property here, it might be less invasive to introduce an intermediary `@not_discoverable` class in the hierarchy to represent the old inheritance tree. Something like:

```py
@not_discoverable 
class MSSQLTableBase(MSSQLBase):
    table = StringProperty(title='Table', default='{{ $table }}', order=5)
```

Then have the existing blocks inherit from that... there might be other items that should be extracted into that helper as well, _i.e._ `validate_column()`. That should reduce the overall `diff` in this PR.

Aside; there is an interesting `self.cursor` here that I haven't figured out what it's supposed to do yet... It may just be junk code.

File: mssql_insert_block.py [hunk 2]
```diff
@@ -1,12 +1,13 @@
...
 from .mssql_base import MSSQLBase
 from nio.block.mixins.enrich.enrich_signals import EnrichSignals
 
 
 class MSSQLInsert(EnrichSignals, MSSQLBase):
 
-    version = VersionProperty("0.2.0")
-    row = Property(title='Row', default='{{ $.to_dict() }}')
+    version = VersionProperty("1.0.0")
+    table = StringProperty(title='Table', default='{{ $table }}', order=5)
```
Comment (32bitkid) [inline mssql_insert_block.py]: i feel like these numberings are all off since a lot of stuff has been extracted into the `Connection` property holder.

File: mssql_rawquery_block.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            output_signals = []
+
+            for signal in signals:
+                _query = self.query(signal)
+                _parameters = self.parameters(signal)
+
+                result = cursor.execute(_query, _parameters) if len(_parameters) > 0 else cursor.execute(_query)
+
+                try:
+                    rows = result.fetchall()
```
Comment (32bitkid) [inline mssql_rawquery_block.py]: this is more of an interesting/philosophical question: given the result-set is _huge_, is doing a `fetchall()` memory safe? or should it self partition into multiple [`fetchmany()`](https://www.python.org/dev/peps/pep-0249/#fetchmany) and dispatch them in chunks.

File: mssql_rawquery_block.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+from nio.block.terminals import output
+from .mssql_base import MSSQLBase
+from nio.block.mixins.enrich.enrich_signals import EnrichSignals
+
+@output('results', label='Results')
+@output('no_results', label='No Results')
+class MSSQLRawQuery(EnrichSignals, MSSQLBase):
+
+    version = VersionProperty("1.0.0")
+    query = StringProperty(title='Parameterized Query (use ? for any user-supplied values)', default='SELECT * FROM table where id=?', order=10)
```
Comment (32bitkid) [inline mssql_rawquery_block.py]: it might be better to use one of the alternate [paramstyle](https://www.python.org/dev/peps/pep-0249/#paramstyle) options: namely `numeric`. In the case where you want to reuse a single parameter multiple times:

```sql
SELECT baz FROM tbl where foo=:1 and bar=:1
```

or perhaps the ability to select which paramstyle makes the most sense

File: mssql_conditions.py [hunk 5]
```diff
@@ -47,9 +32,7 @@ def get_where_conditions(self, signal, table, cursor):
             else:
                 conditions += ' {} '.format(combine_condition)
 
-            condition_string = '{} {} ?'.format(
-                self.validate_column(condition.column(signal), table, cursor),
-                condition.operation(signal).value)
+            condition_string = '{} {} ?'.format(self.validate_column(condition.column(signal), table, cursor), condition.operation(signal).value)
```
Comment (32bitkid) [inline mssql_conditions.py]: i know this was here before, but this string formatting stuff makes me _really_ uncomfortable, especially when it is derived from incoming signals. I, literally, cannot express how uncomfortable it makes me.

[Bobby Tables](https://xkcd.com/327/) does not approve.

File: spec.json [hunk 6]
```diff
@@ -1,289 +1,314 @@
-{
-  "nio/MSSQLDelete": {
-    "version": "0.2.0",
-    "description": "Creates a cursor object for an incoming signal and executes one command per signal.",
-    "categories": [
-      "Database"
-    ],
-    "properties": {
-      "command": {
```
Comment (32bitkid) [inline spec.json:9]: i think there is a way to derive this stuff from python reflection. See https://github.com/nio-blocks/communication/blob/master/spec.json

File: release.json [hunk 7]
```diff
@@ -1,22 +1,27 @@
...
+{
+  "nio/MSSQLRawQuery": {
+    "language": "Python",
+    "url": "git://github.com/nio-blocks/mssql.git",
+    "version": "1.0.0"
+  },
+  "nio/MSSQLDelete": {
+    "language": "Python",
+    "url": "git://github.com/nio-blocks/mssql.git",
+    "version": "1.0.0"
```
Comment (32bitkid) [inline release.json]: Python reflection: https://github.com/nio-blocks/communication/blob/master/release.json#L9

File: mssql_base.py [hunk 8]
```diff
@@ -1,91 +1,85 @@
...
+    port = IntProperty(title='Port', default='[[MSSQL_PORT]]', order=2)
+    database = StringProperty(title='Database', default='[[MSSQL_DB]]', order=3)
+    userid = StringProperty(title='User ID', allow_none=True, default='[[MSSQL_USER]]', order=4)
+    password = StringProperty(title="Password", allow_none=True, default='[[MSSQL_PWD]]', order=5)
+    mars = BoolProperty(title='Enable Multiple Active Result Sets', default=False, order=6)
+
+@not_discoverable
+class MSSQLBase(Block):
+
+    version = VersionProperty('1.0.0')
```
Comment (32bitkid) [inline mssql_base.py]: @mattdodge does a `@not_discoverable` block need a version?

Comment (shadetree01010100) [reply mssql_base.py]: Obviously I'm not Matt, but I think it does. One example of "why" might be that it has configurable Properties, so plausibly not all configurations are compatible with all revisions of the code.

Comment (shadetree01010100) [reply mssql_base.py]: In this case I can find precedence: https://github.com/nio-blocks/state/blob/master/state_base.py#L27

Comment (32bitkid) [reply mssql_base.py]: well, what I mean, is that you cannot create/discover an instance of this block by itself—I suppose in OO parlance, it's more or less an `abstract` class. In this particular case, the version property on this base class is going to get more or less obliterated by the inheriting classes, so I was wondering if it's even necessary.

Comment (shadetree01010100) [reply mssql_base.py]: Oh yea that's a good point.

File: mssql_rawquery_block.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    parameters = Property(title='Substitution Parameters (As List, In Order)', default='{{[]}}', order=11)
```
Comment (32bitkid) [inline mssql_rawquery_block.py]: I'd be really tempted to use a `ListProperty(Property)` here, rather than an expression that evaluates to a list

File: mssql_rawquery_block.py (hunk 4 above)
Comment (32bitkid) [reply mssql_rawquery_block.py]: also here, is it possible to define a, what I called in the kotlin and javascript stacks, `StaticProperty`? A property that is guaranteed not to be interpolated and you just get the raw string value, or whatever the underlying JSON type is in the config. Evaluating `_query = self.query(signal)` opens it up to the same signal/SQL injection issue, by-passing the parametric safety that you get with `.execute(query, params)`.

at the least, it would be safer to evaluate this outside of a signal context—I wish there was a better indicator of this that is propagated outward via the API so you know if this is a signal-property or a block-property—by just doing `self.query()` outside the loop or on `start()`

File: mssql_conditions.py (hunk 5 above)
Comment (32bitkid) [reply mssql_conditions.py]: Actually, I parsed this code a little closer, and it's not as bad as I thought it was, most of these blocks get pushed through `get_where_conditions()` which will call `validate_column()`, which at least checks that the column exists on the table. And then uses parameter injection. Unfortunately, the `table` names are injected directly, and evaluated from the signal. So `{ "table": "table; DROP TABLE Users; --" }` is going to be less than ideal. It probably wouldn't be outrageous to do the same column-check on tables as well—or treat table like a static/block property rather than a signal-property, so it's at least not injectable. _I.e._ if you screw up writing the expression in your _service_ then thats on you.

However, there one block that _does_ still seem dangerous is the InsertBlock. That one doesn't seem to do any validation and is all string injection, with its _own_ string escaping! Yikes.

File: mssql_rawquery_block.py (hunk 3 above)
Comment (mattdodge) [reply mssql_rawquery_block.py]: Assuming they can use the `LIMIT` operator in the SQL statement I'm ok with `fetchall()` for now. It kind of leaves the pagination/cursoring up to the block configurer. I wouldn't be opposed to a "paged" block or block property though that took a cursor/page size and did this for the user.

File: mssql_rawquery_block.py (hunk 4 above)
Comment (mattdodge) [reply mssql_rawquery_block.py]: I really thought we had a property argument called `allow_expr` that could be set to False for this kind of thing. The Retry mixin [actually uses it](https://github.com/niolabs/nio/blob/6326f08f59c33671dc522116562f3543d8396cca/nio/block/mixins/retry/retry.py#L21). But it doesn't seem to actually exist...

File: mssql_base.py (hunk 8 above)
Comment (mattdodge) [reply mssql_base.py]: It does not need a version property but if it has one the version will be inherited by subclassed blocks. Typically we won't version parent blocks though, that allows a sub block to version bump without other sub blocks doing so. In this case I would think we'd want version props only on the sub blocks.

File: README.md [hunk 10]
```diff
@@ -1,133 +1,151 @@
...
+
+MSSQLRawQuery
+=============
+Creates a cursor object for every list of signals, and executes one query per signal. Changes are committed automatically for each list of signals. When changes are committed, all active cursors (i.e. lists of signals) for that connection (i.e. instance of the block) are committed simultaneously. To avoid this behavior use multiple instances of the block.
+
+Properties
+----------
+- **connection**: userid, password, server, port, database, mars
+- **enrich**: enable signal enrichment
+- **id**: block id
```
Comment (mattdodge) [inline README.md]: Don't need to document the ID field either

File: README.md [hunk 11]
```diff
@@ -1,133 +1,151 @@
...
+
+***
+
+MSSQLRawQuery
+=============
+Creates a cursor object for every list of signals, and executes one query per signal. Changes are committed automatically for each list of signals. When changes are committed, all active cursors (i.e. lists of signals) for that connection (i.e. instance of the block) are committed simultaneously. To avoid this behavior use multiple instances of the block.
+
+Properties
+----------
+- **connection**: userid, password, server, port, database, mars
```
Comment (mattdodge) [inline README.md]: In the README the property bullet points should reflect the "title" or label of the property, not the property name. This is because these docs are visible on the designer next to the block edit modal. In other words, use "Database Connection" here rather than `connection`.

File: mssql_base.py [hunk 12]
```diff
@@ -1,91 +1,85 @@
...
+    connection = ObjectProperty(Connection, title='Database Connection', order=1, advanced=True)
```
Comment (mattdodge) [inline mssql_base.py]: Even though some of the properties are rather advanced (i.e. MARS), the connection object shouldn't be advanced. You can't realistically use these blocks without the connection, so it should be visible all the time.

I'd propose keeping the connection object non-advanced and moving MARS to an advanced property, outside the connection object.

File: mssql_rawquery_block.py (hunk 9 above)
Comment (32bitkid) [reply mssql_rawquery_block.py]: i guess the only reason that I can think of where you _would_ want to do this is if you are dynamically selecting the query based on a signal, which has different length parameter sets. To me, thats an *anti*-feature, though something that is _allowed_ based on the defined API. Something like:

```
query: "{{ 'select * from x where foo=?' if $foo is 'bar' else 'delete from x where bar=? and baz=?' }}"
parameters: "{{ [$wat] if $foo is 'bar' else [$bar, $baz] }}"
````

But *sweet jesus*, don't do that.

File: mssql_rawquery_block.py [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+                cursor = self.cnxn.cursor()
+            except Exception as e:
+                self.disconnect()
+                self.connect()
+                cursor = self.cnxn.cursor()
+
+            output_signals = []
+
+            for signal in signals:
+                _query = self.query(signal)
```
Comment (mattdodge) [inline mssql_rawquery_block.py]: I don't think we can call `self.query` in any sort of dynamic signal-based context like this. There's just no way that's ever going to be up to snuff with security. If they are using the raw query block, they have to put a static query in here. The query parameters are where we can use an incoming signal's attributes.

Basically, this line needs to change to
```python
_query = self.query()
```

Comment (mattdodge) [reply mssql_rawquery_block.py]: The same is true for `self.table` elsewhere in this code. Table names should be fixed and not dependent on incoming signal attributes - the example @32bitkid gave earlier is the reason why.

File: mssql_rawquery_block.py [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+                except Exception:
+                    output_signals.append(self.get_output_signal({'inserted': result.rowcount}, signal))
+
+            cursor.commit()
+            cursor.close()
+
+            if len(output_signals) > 0:
+                self.notify_signals(output_signals, output_id='results')
+            else:
+                output_signals.append(self.get_output_signal({'results': 'null'}, signals[0]))
```
Comment (mattdodge) [inline mssql_rawquery_block.py]: Is there a reason we use the string null here instead of the Python null (`None`)?

File: tests/test_mssql_rawquery_block.py [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+    @patch(MSSQLBase.__module__ + '.pyodbc')
+    def test_exception_handling(self, mock_odbc):
+        mock_cnxn = mock_odbc.connect.return_value = MagicMock()
+        mock_cursor = mock_cnxn.cursor.side_effect = Exception()
+        blk = MSSQLRawQuery()
+        self.configure_block(blk, self.config)
+        blk.start()
+        self.assertEqual(mock_odbc.connect.call_count, 1)
+        with self.assertRaises(Exception):
+            blk.process_signals([Signal({'query': 'SELECT * FROM ? WHERE foo = ? AND pi > ?', 'parameters': ['a_table', 'bar', 3]})])
```
Comment (mattdodge) [inline tests/test_mssql_rawquery_block.py]: This test shows exactly why we shouldn't be processing queries off of signal attributes. A signal containing a SQL query that gets executed is about the scariest thing I can imagine seeing. 😱

**Review by mattdodge — CHANGES_REQUESTED** [review]
Some more comments adding to what is there.

@deliciousmonster can you also remove the execute bit from the files where it got added? Unless executing a README file does something I'm quite unaware of it makes the PR more cluttered and isn't needed.

File: tests/test_mssql_rawquery_block.py (hunk 15 above)
Comment (deliciousmonster) [reply tests/test_mssql_rawquery_block.py]: this has been fixed

File: mssql_rawquery_block.py (hunk 14 above)
Comment (deliciousmonster) [reply mssql_rawquery_block.py]: no, fixed

File: mssql_rawquery_block.py (hunk 13 above)
Comment (deliciousmonster) [reply mssql_rawquery_block.py]: This has been changed to 
`_query = self.query()`

File: README.md (hunk 11 above)
Comment (deliciousmonster) [reply README.md]: fixed

File: README.md (hunk 10 above)
Comment (deliciousmonster) [reply README.md]: removed

File: mssql_rawquery_block.py (hunk 9 above)
Comment (deliciousmonster) [reply mssql_rawquery_block.py]: Converted to ListProperty[Property]

File: release.json (hunk 7 above)
Comment (deliciousmonster) [reply release.json]: done. that's neat.

File: mssql_rawquery_block.py (hunk 3 above)
Comment (deliciousmonster) [reply mssql_rawquery_block.py]: In the service implementation, I've set an env var to limit GETs to 1000 unless otherwise specified

File: tests/test_mssql_rawquery_block.py (hunk 15 above)
Comment (32bitkid) [reply tests/test_mssql_rawquery_block.py]: tangentally-related to this, I don't like that [this signal](https://github.com/nio-blocks/mssql/blob/master/mssql_update_block.py#L58) generated in the update block uses the key of `"Rows updated"`. It just makes it annoying to get out a value, whether it's python or javascript.

File: mssql_base.py [hunk 16]
```diff
@@ -1,91 +1,105 @@
...
+    user_id = StringProperty(title='User ID', allow_none=True, default='[[MSSQL_USER]]', order=4)
+    password = StringProperty(title='Password', allow_none=True, default='[[MSSQL_PWD]]', order=5)
+
+class Mars(PropertyHolder):
+    enabled = BoolProperty(title='Enable MARS?', default=True, order=1)
+
+@not_discoverable
+class MSSQLBase(Block):
+    connection = ObjectProperty(Connection, title='Database Connection', order=50)
+    mars = ObjectProperty(Mars, title='Multiple Active Result Sets (MARS)', order=60)
```
Comment (mattdodge) [inline mssql_base.py:20]: I don't think this needs to be an object property (it only has one object) and it should be advanced. Think this can be replaced with
```python
mars = BoolProperty(title='Enable MARS?', default=True, advanced=True)
```

File: mssql_base.py [hunk 17]
```diff
@@ -1,91 +1,105 @@
...
+        try:
+            return self.cnxn.cursor()
+        except Exception as e:
+            self.disconnect()
+            self.connect()
+            return self.cnxn.cursor()
+
+
+@not_discoverable
+class MSSQLTabledBase(MSSQLBase):
```
Comment (mattdodge) [inline mssql_base.py]: This is a good idea for a new class, thanks. Please move it to a separate file though `mssql_table.py` - this is a style guideline thing.

File: mssql_query_block.py [hunk 18]
```diff
@@ -1,53 +1,49 @@
...
-
-    def process_signals(self, signals):
-        if self.isConnecting:
-            self.logger.error(
-                'Connection already in progress. Dropping signals.')
+class MSSQLQuery(EnrichSignals, MSSQLTabledBase, MSSQLConditions):
+
+    version = VersionProperty("1.0.1")
+
+    def process_signals(self, signals, **kwargs):
```
Comment (mattdodge) [inline mssql_query_block.py]: Just curious, why did you add `**kwargs` here?

Comment (mattdodge) [reply mssql_query_block.py]: And it looks like in other process_signals calls too

**Review by mattdodge — CHANGES_REQUESTED** [review]
This is looking really good and pretty close I think. A couple of quick requests then I think we're good to go!

File: mssql_query_block.py (hunk 18 above)
Comment (32bitkid) [reply mssql_query_block.py]: This probably me and probably a PyCharm Pro auto-fix/auto-formatting to match the signature type of the overridden member; the parent signature is 

```
def process_signals(self, signals, input_id=DEFAULT_TERMINAL):
```

So, probably overkill and I missed it on my clean-up, since its over-ridden completely it doesn't really matter.

File: mssql_base.py (hunk 16 above)
Comment (deliciousmonster) [reply mssql_base.py:20]: I put that in there because it looks nicer when it's wrapped in a fieldset... and I didn't hide it because it's a pretty major settin, at least for the RawQuery... so then I didn't want to hide it for some blocks, but not others...

File: mssql_query_block.py (hunk 18 above)
Comment (deliciousmonster) [reply mssql_query_block.py]: I removed them... just self, signals now

File: mssql_base.py (hunk 16 above)
Comment (mattdodge) [reply mssql_base.py:20]: Haha well if it's not advanced I guess we could have just left it in the connection properties then. Seems like an advanced property to me but I'll take your word for it since you have used the blocks more 👍

**Review by mattdodge — APPROVED** [review]
